contracts/minters/base-minter/src/contract.rs

```diff
@@ -107,61 +111,146 @@ pub fn execute(
+            // Create network fee msgs
+            let mint_fee_percent = factory_params.mint_fee_bps.bps_to_decimal();
+            let network_fee = match mint_price {
+                sg2::Token::Fungible(coin) => coin.amount * mint_fee_percent,
+                sg2::Token::NonFungible(_) => Uint128::new(0),
+            };
+            // TODO: NFTs don't have a fee
+            // For the base 1/1 minter, the entire mint price should be Fair Burned
+            if network_fee != funds_sent {
+                return Err(ContractError::InvalidMintPrice {});
+            }
+            checked_fair_burn(&info, network_fee.u128(), None, &mut res)?;
+            Ok(network_fee)
+        }
+    }
+}
+
+fn _execute_mint_sender(
+    deps: DepsMut,
+    env: Env,
     info: MessageInfo,
     token_uri: String,
 ) -> Result<Response, ContractError> {
     let config = CONFIG.load(deps.storage)?;
-    let collection_address = COLLECTION_ADDRESS.load(deps.storage)?;
 
+    let collection_address = COLLECTION_ADDRESS.load(deps.storage)?;
     // This is a 1:1 minter, minted at min_mint_price
     // Should mint and then list on the marketplace for secondary sales
     let collection_info: CollectionInfoResponse = deps.querier.query_wasm_smart(
         collection_address.clone(),
         &Sg721QueryMsg::CollectionInfo {},
     )?;
-    // allow only sg721 creator address to mint
-    if collection_info.creator != info.sender {
-        return Err(ContractError::Unauthorized(
-            "Sender is not sg721 creator".to_owned(),
-        ));
-    };
+
+    let _ = _check_sender_is_collection_or_contract(
+        env.contract.address.clone(),
+        info.clone(),
+        collection_info.clone(),
+    )?;
 
     let parsed_token_uri = Url::parse(&token_uri)?;
     if parsed_token_uri.scheme() != "ipfs" {
         return Err(ContractError::InvalidTokenURI {});
     }
 
     let mut res = Response::new();
-
     let factory: ParamsResponse = deps
         .querier
         .query_wasm_smart(config.factory, &Sg2QueryMsg::Params {})?;
     let factory_params = factory.params;
 
-    let funds_sent = must_pay(&info, NATIVE_DENOM)?;
-
-    // Create network fee msgs
-    let mint_fee_percent = factory_params.mint_fee_bps.bps_to_decimal();
-    let network_fee = config.mint_price.amount * mint_fee_percent;
-    // For the base 1/1 minter, the entire mint price should be Fair Burned
-    if network_fee != funds_sent {
-        return Err(ContractError::InvalidMintPrice {});
-    }
-    checked_fair_burn(&info, network_fee.u128(), None, &mut res)?;
-
+    let network_fee = _pay_mint_if_not_contract(
+        env.contract.address,
+        info.clone(),
+        config.mint_price,
+        factory_params,
+    )?;
     // Create mint msgs
     let mint_msg = Sg721ExecuteMsg::<Extension, Empty>::Mint {
         token_id: increment_token_index(deps.storage)?.to_string(),
-        owner: info.sender.to_string(),
+        owner: collection_info.creator,
```

Comment:
> the NFT owner would be the collection creator?

So now that the contract can call itself through the execute_burn_to_mint messages, we can't just assume the the sender is always creator, and therefore info.sender won't cut it.

Actually looking at the name of this function `execute_mint_sender` I'm not entirely sure what is the intention of this function. It sounds like a mint is being performed on behalf of the sender? But here I am just putting back the collection creator because that's all I have access to to figure out who should be the owner. Unless I do a query for owner here? Thoughts?